### ISSUE TITLE
Support for Shenandoah GC

### ISSUE THREAD

**Issue by rkennke** [oracle/graal#3472]
The Graal compiler currently does not support the Hotspot Shenandoah GC. Shenandoah GC is relevant in situations where latency is a concern, because it drastically reduces GC pause times. This is not only relevant in large-heap settings where GC tends to take a long time due to the size of live data that needs to be marked and relocated, but also in resource-constrained environment where even relatively little STW-work can balloon.

A first step in supporting Shenandoah GC in Graal compiler would be to implement the basic Shenandoah barriers:
- Load-reference barrier to be employed after loading references
- SATB-barrier to be employed before reference writes (very similar to G1's SATB barrier)
- CAS-obj barrier that handle Unsafe.compareAndSwapReference() and related methods
- Arraycopy barrier to deal with arraycopies of Object[]
- Reference.get() barrier

This would be enough to support Graal compiler in JDK 11. However, additional features have been added between JDK 11 and JDK 17, which require extra work:
- Concurrent class-unloading requires so-called nmethod-barriers: little pieces of GC code to be called when calling into 'armed' nmethods (and a mechanism to arm nmethods).
- Concurrent thread-scanning which requires stack-watermark support.

Both these features are not Shenandoah specific. For example, ZGC would also make use of them.

Much of the first step (JDK 11 support) is implemented in #2426.

**Comment by mcraj017** [comment]
@dougxc please advise

**Comment by rkennke** [comment]
Notice that this is mostly informational for now. We recently had a discussion with @thomaswue were we agreed that I file this issue to collect TODOs and a rough outline, also with an eye on possibly supporting Shenandoah in native-image.

**Comment by dougxc** [comment]
I'm closing this issue due to lack of activity. Please re-open it (and reassign) when/if there are plans to make progress on adding Shenandoah to Native Image.

**Comment by bric3** [comment]
@dougxc I don't think this issue is related to Shenandoah on the native image but on the Graal Compiler.

**Comment by reneleonhardt** [comment]
It's sad that there is no progress after 3 years, switching from G1 to Shenandoah reduced our spring-web latency massively.

**Comment by JohnTortugo** [comment]
Please, re-open this issue. I'm resuming @rkennke work on this patch.

**Comment by rkennke** [comment]
I cannot seem to assign this back to me. I'm currently working on it, WIP PR is here: https://github.com/oracle/graal/pull/10904

**Comment by zakkak** [comment]
Closed by https://github.com/oracle/graal/pull/10904 (see https://github.com/oracle/graal/commit/3c7b5cb092a606d2d9c421dbe225bff16789fb2d)

Native-image Shenandoah support tracked in https://github.com/oracle/graal/issues/12237

**Comment by Venorcis** [comment]
> Closed by https://github.com/oracle/graal/pull/10904 (see https://github.com/oracle/graal/commit/3c7b5cb092a606d2d9c421dbe225bff16789fb2d)

Great news 🚀 Is this planned to be included in an upcoming release as well? As it's not available in the current 25.0.1 yet.

**Comment by thomaswue** [comment]
Yes. Specifically it is planned to be available for native image. See this ticket: https://github.com/oracle/graal/issues/12237

**Comment by Venorcis** [comment]
> Yes. Specifically it is planned to be available for native image. See this ticket: [#12237](https://github.com/oracle/graal/issues/12237)

Thank you for the quick response! Will it also be available without native image? As on GraalVM (CE) I get `jdk.graal.compiler.debug.GraalError: Shenandoah garbage collector is not supported by Graal` atm.
